File: OEM/Domain/IncidentTypes/IncidentTypeCode.cs
namespace OEMAPI.Domain.IncidentTypes
{
    public class IncidentTypeCode : ValueObject
    {
        public string Value { get; }

        public IncidentTypeCode(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
                throw new BusinessRuleValidationException("Incident type code cannot be empty.");

            Value = value;
        }

        protected override IEnumerable<object> GetEqualityComponents()
        {
            yield return Value;
        }
    }
}

File: OEM/src/Domain/IncidentTypes/IncidentTypeCode.ts
import { ValueObject } from "../../core/domain/ValueObject";
import { Result } from "../../core/logic/Result";
import { Guard } from "../../core/logic/Guard";

interface IncidentTypeCodeProps {
  value: string;
}

export class IncidentTypeCode extends ValueObject<IncidentTypeCodeProps> {

  get value(): string {
    return this.props.value;
  }

  private constructor(props: IncidentTypeCodeProps) {
    super(props);
  }

  public static create(code: string): Result<IncidentTypeCode> {
    const guardResult = Guard.againstNullOrUndefined(code, "code");

    if (!guardResult.succeeded) {
      return Result.fail<IncidentTypeCode>(guardResult.message);
    }

    return Result.ok<IncidentTypeCode>(
      new IncidentTypeCode({ value: code })
    );
  }
}
